### PR TITLE
Symmetry christ

### LINKED ISSUE #299: Leverage Symmetry in Christoffel Symbols to improve the performance of symbolic module
🐞 **Problem**
  - Christoffel Symbols are symmetric over 2nd and 3rd indices

🎯 **Goal**

  - Use this property to speed up it's calculation from metric tensor

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #<put issue number here>`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

### PR DESCRIPTION
**Pull request by OVSofia**
Fixes #299

### CODE REVIEW

**Comment by OVSofia** [comment]
I have to wait for einstein merge to remove the other commits.
I thought to do it differently. But if this is OK, I leave it that way, since the code would be less intuitive doing it as I thought.

**Comment by OVSofia** [comment]
Anyway my approach:
Previously calculate the positions of t at which the position would be repeated. But the `if` would be different depending on the value of dims, so it would be a bit ugly this way xD
```
        for t in range(dims ** 3):
            if (0< t < dims-1) OR (dims+1 < t < 2dims) OR (2dims +2 < t < 3dims) OR etc etc 
               continue
            else:
               k = t % dims
                j = (int(t / dims)) % (dims)
                i = (int(t / (dims ** 2))) % (dims)
                tmpvar = 0
                for n in range(dims):
                    tmpvar += (matinv[i, n] / 2) * (
                        sympy.diff(mat[n, j], syms[k])
                        + sympy.diff(mat[n, k], syms[j])
                        - sympy.diff(mat[j, k], syms[n])
                    )
                tmplist[i][j][k] = tmplist[i][k][j] = tmpvar
        return cls(tmplist, syms, config="ull", parent_metric=metric)
```

**Comment by ritzvik** [comment]
Would review extensively, we would first complete this before going to #303 . OK?

**Comment by ritzvik** [comment]
you can rebase!

File: src/einsteinpy/symbolic/christoffel.py [hunk 1]
```diff
@@ -78,13 +78,16 @@ def from_metric(cls, metric):
             j = (int(t / dims)) % (dims)
             i = (int(t / (dims ** 2))) % (dims)
             tmpvar = 0
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/christoffel.py]: move this `tmpvar` inside the `k<=j` condition

File: src/einsteinpy/symbolic/christoffel.py [hunk 2]
```diff
@@ -78,13 +78,16 @@ def from_metric(cls, metric):
...
+            if k <= j:
+                for n in range(dims):
+                    tmpvar += (matinv[i, n] / 2) * (
+                        sympy.diff(mat[n, j], syms[k])
+                        + sympy.diff(mat[n, k], syms[j])
+                        - sympy.diff(mat[j, k], syms[n])
+                    )
+                tmplist[i][j][k] = tmplist[i][k][j] = tmpvar
+            else:
+                continue
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/christoffel.py]: this `else` was not required

File: src/einsteinpy/symbolic/christoffel.py [hunk 3]
```diff
@@ -77,14 +77,16 @@ def from_metric(cls, metric):
...
+            if k <= j:
+                tmpvar = 0
+                for n in range(dims):
+                    tmpvar += (matinv[i, n] / 2) * (
+                        sympy.diff(mat[n, j], syms[k])
+                        + sympy.diff(mat[n, k], syms[j])
+                        - sympy.diff(mat[j, k], syms[n])
+                    )
+                tmplist[i][j][k] = tmplist[i][k][j] = tmpvar
+            continue
```
Comment (ritzvik) [inline src/einsteinpy/symbolic/christoffel.py]: even this continue was not required!!

**Comment by ritzvik** [comment]
> Anyway my approach:
> Previously calculate the positions of t at which the position would be repeated. But the `if` would be different depending on the value of dims, so it would be a bit ugly this way xD
> 
> @ritzvik What do you think? Is it necessary to make the issue like this?
> 
> ```
>         for t in range(dims ** 3):
>             if (0< t < dims-1) OR (dims+1 < t < 2dims) OR (2dims +2 < t < 3dims) OR etc etc 
>                continue
>             else:
>                k = t % dims
>                 j = (int(t / dims)) % (dims)
>                 i = (int(t / (dims ** 2))) % (dims)
>                 tmpvar = 0
>                 for n in range(dims):
>                     tmpvar += (matinv[i, n] / 2) * (
>                         sympy.diff(mat[n, j], syms[k])
>                         + sympy.diff(mat[n, k], syms[j])
>                         - sympy.diff(mat[j, k], syms[n])
>                     )
>                 tmplist[i][j][k] = tmplist[i][k][j] = tmpvar
>         return cls(tmplist, syms, config="ull", parent_metric=metric)
> ```

Didn't understand this. Can you elaborate? Anyways, the current approach seems fine.

**Comment by ritzvik** [comment]
Could you squash em, this time?

**Comment by OVSofia** [comment]
For dims=2: Excluding:  if (0< t <= dims-1) OR (dims+1 < t < 2dims)  OR (2dims +1 < t <= 3dims), without: t=1; t=6

image

For dims=3: Excluding  if (0< t <= dims-1) OR (dims+1 < t < 2dims) , without t = 1,2; t=5; t=10,11; ... 
image

Etc etc...

Get the expression of the if by induction. But in reality it is very long the "generic expresion"... so better not xD 
but perhaps for other cases this approach may interest

**Comment by ritzvik** [comment]
@OVSofia , still couldn't understand what you are tying to say. Maybe you can explain this on chat sometime.

Anyways, the PR looks good
cc @shreyasbapat 
We can increase the cognitive complexity in codeclimate YAML file, it's too annoying

Can merge!

**Comment by Varunvaruns9** [comment]
> We can increase the cognitive complexity in codeclimate YAML file, it's too annoying

I had to refactor twice to remove the errors 😢

**Comment by ritzvik** [comment]
> > We can increase the cognitive complexity in codeclimate YAML file, it's too annoying
> 
> I had to refactor twice to remove the errors 😢

Increased the threshold

**Pull request merged**